tests/std/tests/P2502R2_generator_promise/test.cpp

```diff
@@ -0,0 +1,242 @@
+    static_assert(semiregular<Promise>);
+
+    Promise p;
+
+    // Test that operator& for promise_type resolves to the built-in version and doesn't involve ADL
+    static_assert(same_as<decltype(&p), Promise*>);
+    assert(&p == addressof(p));
+
+    // Test 'get_return_object'
+    static_assert(same_as<decltype(p.get_return_object()), Gen>);
+    static_assert(noexcept(p.get_return_object()));
+
+    // Test 'initial_suspend'
+    static_assert(same_as<decltype(p.initial_suspend()), suspend_always>);
+    static_assert(same_as<decltype(as_const(p).initial_suspend()), suspend_always>);
+    static_assert(noexcept(p.initial_suspend()));
+    static_assert(noexcept(as_const(p).initial_suspend()));
+
+    // Test 'final_suspend'
+    using FinalAwaitable = decltype(p.final_suspend());
+    static_assert(convertible_to<decltype(declval<FinalAwaitable&>().await_ready()), bool>);
+    static_assert(noexcept(p.final_suspend()));
+
+    // Test 'yield_value(yielded)'
+    static_assert(same_as<decltype(p.yield_value(declval<Yielded>())), suspend_always>);
+    static_assert(noexcept(p.yield_value(declval<Yielded>())));
+
+    // Test 'yield_value(const remove_reference_t<yielded>&)'
+    if constexpr (is_rvalue_reference_v<Yielded>
+                  && constructible_from<remove_cvref_t<Yielded>, const remove_reference_t<Yielded>&>) {
+        using Lval      = const remove_reference_t<Yielded>&;
+        using Awaitable = decltype(p.yield_value(declval<Lval>()));
+
+        static_assert(convertible_to<decltype(declval<Awaitable&>().await_ready()), bool>);
+        static_assert(is_void_v<decltype(declval<Awaitable&>().await_resume())>);
+        static_assert(noexcept(p.yield_value(declval<Lval>()))
+                      == is_nothrow_constructible_v<remove_cvref_t<Yielded>,
+                          const remove_reference_t<Yielded>&>); // strengthened
+    }
+
+    { // Test 'yield_value(elements_of<generator>)'
+        {
+            using Awaitable = decltype(p.yield_value(ranges::elements_of{declval<Gen&>()}));
+            static_assert(convertible_to<decltype(declval<Awaitable&>().await_ready()), bool>);
+        }
+
+        if constexpr (!is_void_v<Alloc>) {
+            using Awaitable = decltype(p.yield_value(ranges::elements_of{declval<Gen&>(), declval<Alloc&>()}));
+            static_assert(convertible_to<decltype(declval<Awaitable&>().await_ready()), bool>);
+        }
+    }
+
+    using ValTy = conditional_t<is_void_v<V>, remove_cvref_t<Ref>, V>;
+    if constexpr (!TestingIncomplete && convertible_to<ValTy&, Yielded>) {
+        // Test 'yield_value(ranges::elements_of<range>)'
+        test_yield_elements_of_range<Gen, vector<ValTy>>(p);
+        test_yield_elements_of_range<Gen, list<ValTy>>(p);
+        test_yield_elements_of_range<Gen, forward_list<ValTy>>(p);
+        test_yield_elements_of_range<Gen,
+            test::range<input_iterator_tag, ValTy, test::Sized::no, test::CanDifference::no, test::Common::no,

```

Comment:
`range_algorithm_support.hpp` drags it in, but we should probably include `<iterator>` for this direct mention of `input_iterator_tag`.

Reply:
Time for my regular "It's Standard that `<ranges>` includes `<iterator>` includes `<concepts>`" reminder. On a related note, I can't wait to start writing `import std;` in tests.